lib/phil/charlie_cards/importer.ex

```diff
@@ -0,0 +1,61 @@
+defmodule Phil.CharlieCards.Importer do
+  @moduledoc """
+  Imports CharlieCard inventory into Phil.
+  """
+
+  import Ecto.Query, only: [from: 2]
+
+  alias NimbleCSV.RFC4180, as: CSV
+
+  alias Phil.CharlieCards.CharlieCard
+  alias Phil.Products.Product
+  alias Phil.Repo
+
+  @media %{"2" => :ticket, "5" => :card}
+
+  @doc """
+  Given the path of a file, this function attempts to import CharlieCards.
+
+  It returns a map of successes and failures like %{error: [], ok: []}.
+  Error entries are changesets and success entries are the inserted struct.
+  """
+  def import(path) do
+    path
+    |> File.stream!()
+    |> CSV.parse_stream()
+    |> Stream.map(&row_to_changeset/1)
+    |> Stream.chunk_every(100)
+    |> Task.async_stream(&insert_batch/1, max_concurrency: 10, ordered: false)
+    |> Enum.reduce([], fn {:ok, results}, acc -> acc ++ results end)
+    |> Enum.group_by(fn {status, _} -> status end, fn {_, value} -> value end)
+  end
+
+  defp insert_batch(changesets) do
+    Enum.map(changesets, fn changeset ->
+      if changeset.valid? do
+        Repo.insert(changeset)
```

Comment:
should we be upserting?

Reply:
I am thinking no because if there are two rows with duplicate information you'll end up with one being imported and you won't know about the problem.

Reply:
that depends on what we use as our unique index. The main use case i can think of for repeated data is someone uploading the same CSV more than once. Have we evidence that data can be repeated _within_ a CSV? What would even happen in that case?

Reply:
They will get told on a row-by-row basis what row is bad and what is wrong with it. So, if they try to import the same CSV twice they'll be told all of those serial numbers are taken. I don't think we should use the import as a bulk edit.

Reply:
Yea I guess I don't understand a lot about how the data will be updated, but with an upsert we don't have to do anything if the record exists already. It's just as a user do I want to see 4k errors because I uploaded the same CSV twice, or do i want to see a success message because at the end of the day the result is the same, which is that those records are definitely imported. 

For example imagine if we have a bug and only half the CSV imports, they'd have to edit the csv to remove rows?  might be confusing

Reply:
No, they could use the same file and the duplicates will just fail as duplicates.